PLANNER-736: Workbench: Prevent user from creating multiple @PlanningSolution data objects

mcimbora opened the pull request:
Introduce pre-save & pre-copy checks to make sure the user doesn't create multiple Planning Solutions within a project. The checks reside on the server side and can be easily defined by implementing PreSaveOperationCheck & PreCopyOperationCheck classes.

DataModelerOperationCheckService provides methods to verify all conditions (checks) in a single place.

As the checks are performed on the server side, there was a need to provide mapping between the server-side messages & client-side messages (where the latter are i18ned). This is managed by OperationCheckTranslator class.

Part of:
https://github.com/droolsjbpm/guvnor/pull/416
https://github.com/droolsjbpm/jbpm-wb/pull/630
https://github.com/droolsjbpm/kie-wb-common/pull/681
https://github.com/droolsjbpm/optaplanner-wb/pull/134

@manstis / @wmedvede Can you please have a look at the PR? Thanks.

mcimbora commented:
The test failures are unrelated.

manstis commented:
@mcimbora Test failures may be unrelated; but this PR will need to be greed before it's merged. You may need to track down a culprit to fix the failing tests.

manstis commented on kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-api/src/main/java/org/kie/workbench/common/screens/datamodeller/operationcheck/ClientSideOperationCheckMessage.java:
```diff
@@ -0,0 +1,67 @@
+    @Override
+    public int hashCode() {
+        int result = source != null ? source.hashCode() : 0;
+        result = 31 * result + ( message != null ? message.hashCode() : 0 );
```
HashCodes should ensure the ```int``` does not overflow in JavaScript land.. Normally we use ```result = ~~result``` to keep it in bounds.

wmedvede commented:
Hello @mcimbora 
I know that when we chatted about this I've informally used the term "check" or "save conditions", I don't remember exaclty to be honest. But now that I'm looking at the PR I believe it could make sence to rename the concept of  "operationscheck" to just validations. I believe that it fits better with different validators we have around the platform.

I'm not adding all the list here but here are some sugestions I believe can make sense.

operationscheck  -> validation
PreCopyOperationCheck -> CopyValidator
PreSaveOperationCheck -> SaveValidator

ServerSideOperationCheckMessage -> ValidationMessage ? 
(can it make sense here to use the already existing org.guvnor.common.services.shared.validation.model.ValidationMessage as the base class? even if the translators can then make particualr processings for sub-classes?)

ServerSideOperationCheckMessage check( Path dataObjectPath, DataObject dataObject ); change to:

List<ValidationMessage> validate( Path dataObjectPath, DataObject dataObject ); (give the chance that a validator can return N messages)

DataModelerOperationCheckService -> DataModelerValiationService
with methods 
List<ValidationMessage> validateForSave( Path dataObjectPath, DataObject dataObject );
List<ValidationMessage> validateForCopy( Path dataObjectPath, DataObject dataObject );

and so on. wdyt?

wmedvede commented on kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/operationcheck/OperationCheckTranslatorUtils.java:
```diff
@@ -0,0 +1,59 @@
+                .filter( t -> t.accept( serverSideOperationCheckMessage ) )
+                .map( t -> t.translate( serverSideOperationCheckMessage ) )
+                .findFirst()
+                .orElseThrow( () -> new IllegalStateException( "No translation found for " + serverSideOperationCheckMessage ) );
```
If a translator could not be fund I believe that likely we can just show the message returned by server, it's likely better to show the un-translated message and avoid the user to see an exception.

wmedvede commented on kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/operationcheck/OperationCheckPopupFactory.java:
```diff
@@ -0,0 +1,134 @@
+                                   String yesButtonText,
+                                   Command noCommand,
+                                   List<ClientSideOperationCheckMessage> clientSideOperationCheckMessages ) {
+        BaseModal modal = new BaseModal();
```
I believe it might be worth to do some effort here and try to separate this into some more MVP like.

wmedvede replied:
A test is also appreciated.

mcimbora commented:
@wmedvede @manstis Thanks. I agree to change the terminology to "validation" as suggested. I'll let you know when the PR is updated.

mcimbora commented:
@manstis @wmedvede PR updated. I tried to reuse what we already have for validation and moved the check logic to `ValidationService` & reused `ValidationMessage` class. I refactored `ValidationPopup` to follow MVP.

manstis commented on kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/validation/ValidationServiceImpl.java:
```diff
@@ -153,4 +167,23 @@ public boolean validateGAVVersion( final String version ) {
+    @Override
+    public <T> Collection<ValidationMessage> validateForSave( final Path path,
+                                                              final T content ) {
+        return (Collection<ValidationMessage>) saveValidators.stream().filter( v -> v.accept( path ) ).flatMap( c -> c.validate( path,
```
Can we really have ```null``` ```ValidationMessage```? (it looks like you filter the list from validation for nulls)

mcimbora replied:
Good spot. A remnant of the old code.

manstis commented on kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/validation/ValidationServiceImpl.java:
```diff
@@ -153,4 +167,23 @@ public boolean validateGAVVersion( final String version ) {
+    @Override
+    public <T> Collection<ValidationMessage> validateForCopy( final Path path,
+                                                              final T content ) {
+        return (Collection<ValidationMessage>) copyValidators.stream().filter( v -> v.accept( path ) ).flatMap( c -> c.validate( path,
```
Same comment about ```null``` filter.

manstis commented on kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/validation/ValidationServiceImpl.java:
```diff
@@ -153,4 +167,23 @@ public boolean validateGAVVersion( final String version ) {
+
+    @Override
+    public Collection<ValidationMessage> validateForCopy( final Path path ) {
+        return (Collection<ValidationMessage>) copyValidators.stream().filter( v -> v.accept( path ) ).flatMap( c -> c.validate( path ).stream().filter( r -> r != null ) ).collect( Collectors.toList() );
```
Same comment about ```null``` filter.

manstis commented on kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/popups/validation/ValidationMessageTranslatorUtils.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    // Test purposes
+    void setValidationMessageTranslators( List<ValidationMessageTranslator> messageTranslators ) {
```
We do have a ```MockInstance<T>``` class floating around somewhere; but this approach is equally fine.

manstis commented on kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/popups/validation/ValidationPopupViewImpl.html:
```diff
@@ -0,0 +1,26 @@
+    <div class="container-fluid">
+        <table id="validationTable"></table>
+        <div class="modal-footer">
+            <button id="cancelButton"></button>
```
Small detail.. isn't it _customary_ to have "OK" (Yes) and then "Cancel" as opposed to "Cancel" then "OK" (Yes)?

mcimbora replied:
I tried to follow the ordering we use in existing Uberfire popups (e.g. copyPopup/savePopup).

manstis commented on kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/resources/i18n/ValidationConstants.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Validation I18N constants
+ */
+public interface ValidationConstants
```
Not that I think any "standard" has been agreed; but if you're using ```errai-ui``` it'd _possibly_ make sense to use ```TranslationService``` too instead of GWT's i18n mechanism.

mcimbora replied:
+1, changing.

mcimbora commented:
@manstis If there are no further comments, this is ready to go in.

manstis commented:
@mcimbora OK with me.. @wmedvede any comments before it's merged?

wmedvede commented:
Hello @mcimbora, it think it's really good now. Just one last comment from my side. I know that in general to code the tests for KieEditor/BaseEditor based workbench editors is in general not easy, and is usually a fault we have, but it'll much appreciated if a test can be added that emulates the following main use cases, 1) the copy/rename operation is invoked, 2) copy/rename validators are set and return ValidationMessages, 3) we assert that the corresponding validation popup is shown.

mcimbora commented:
@wmedvede @manstis PR updated, added tests for save/copy/rename (DataModelerScreenPresenterTest.java). Thanks for the review

wmedvede commented:
+1, thanks!
